### PR TITLE
Spot instances: Fallback to on-demand when there is no capacity

### PR DESCRIPTION
**Pull request by julienduchesne**
Follow up to #331. When there is no spot capacity (either for spot instances with a bid price or without one), fallback to on-demand instances

### CODE REVIEW

File: src/main/java/hudson/plugins/ec2/SlaveTemplate.java [hunk 1]
```diff
@@ -528,7 +529,16 @@ private void logProvisionInfo(String message) {
...
-    private List<EC2AbstractSlave> provisionOndemand(int number, EnumSet<ProvisionOptions> provisionOptions) throws AmazonClientException, IOException {
+    private List<EC2AbstractSlave> provisionOndemand(int number, EnumSet<ProvisionOptions> provisionOptions)
+            throws IOException {
+        return provisionOndemand(number, provisionOptions, false, false);
+    }
+
+    /**
+     * Provisions an On-demand EC2 slave by launching a new instance or starting a previously-stopped instance.
+     */
+    private List<EC2AbstractSlave> provisionOndemand(int number, EnumSet<ProvisionOptions> provisionOptions, Boolean spotWithoutBidPrice, Boolean fallbackSpotToOndemand)
```
Comment (jvz) [inline src/main/java/hudson/plugins/ec2/SlaveTemplate.java]: Same boolean boxing concerns as in #331.

File: src/main/java/hudson/plugins/ec2/SlaveTemplate.java [hunk 2]
```diff
@@ -929,6 +961,25 @@ private void setupCustomDeviceMapping(List<BlockDeviceMapping> deviceMappings) {
                 }
                 String slaveName = spotInstReq.getSpotInstanceRequestId();
 
+                if (spotConfig.fallbackToOndemand) {
+                    for (int i = 0; i < 2 && spotInstReq.getStatus().getCode().equals("pending-evaluation"); i++) {
```
Comment (jvz) [inline src/main/java/hudson/plugins/ec2/SlaveTemplate.java]: I guess it's not exactly required, but jenkins-core somewhat recently added a `Retrier<T>` class which might be useful here. Might be worth looking at? I'm mostly concerned about `Thread.sleep` calls for testing purposes.

Comment (julienduchesne) [reply src/main/java/hudson/plugins/ec2/SlaveTemplate.java]: hmm, the Retrier is very nice. However, it requires us to bump the jenkins version to at least 2.152. And the Retrier is marked `NoExternalUse` which crashes the build (unless there's a way around that).

Comment (jvz) [reply src/main/java/hudson/plugins/ec2/SlaveTemplate.java]: Oh right. NoExternalUse is for avoiding creating new public APIs that can never change again. Ah well.

**Pull request merged**